kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -387,21 +388,56 @@ private CloseableIterator<ActionWrapper> getNextActionsIter() {
     }
   }
 
-  /** Take input (iterator<T>, boolean) and produce an iterator<T, boolean>. */
+  /**
+   * Take input (iterator<T>, boolean) and produce an iterator<T, boolean>. When the input iterator
+   * has been created from a delta file, this function will try to use the inCommitTimestamp (if
+   * available) from the CommitInfo action as the commit timestamp for the file.
+   */
   private CloseableIterator<ActionWrapper> combine(
       CloseableIterator<ColumnarBatch> fileReadDataIter,
       boolean isFromCheckpoint,
       long version,
       Optional<Long> timestamp) {
+    final List<ColumnarBatch> firstActionList = new ArrayList<>();
+    final int commitInfoOrdinal = deltaReadSchema.indexOf("commitInfo");
+    // For delta files, we want to use the inCommitTimestamp from commitInfo
+    // as the commit timestamp for the file.
+    // Since CommitInfo should be the first action in the delta when inCommitTimestamp is
+    // enabled, we will read the first action and try to extract the timestamp from it.
+    // We also cache the first action so that we can return it in the iterator.
```

Comment:
I wonder if we can just implement a default method `peek` on `CloseableIterator`? Would simplify this code

Reply:
If it's the first action don't we not need to check the entire column vector either? (which is done in `InCommitTimestampUtils.getCommitInfo`)

Reply:
+1 alternative have a wrapper like

public class PeekableIterator<T> implements CloseableIterator<T> {
    private final CloseableIterator<T> delegate;
    private T next;

    public PeekableIterator(CloseableIterator<T> delegate) {
        this.delegate = delegate;
        this.next = delegate.hasNext() ? delegate.next() : null;
    }

This way we don't need to have two iterators.